### DISCUSSION TITLE
Can Timefold be used for creating EV charging plans ?

### DISCUSSION THREAD

**Discussion by rhuss** [TimefoldAI/timefold-solver#1204] [General]
Sorry if this is the wrong place to this question (feel free to close it if not appropriate), but I wonder whether Timefold can be used for calculating an intelligent charging plan for electric vehicles with the following possible constraints:

* Charging should be finished by a given end date (e.g., 6:30 am). This is a hard constraint
* The target SoC (state of charge) should be in a range (e.g., 70 - 80% loaded)
* A dynamic tariff table for the grid power should be used to minimize costs (e.g. when using a provider like Tibber, where the costs change from hour to hour. You get the cost data via an API).
* The tricky part: With a connected PV, charging should only happen when excess energy is produced, which otherwise would be sent into the grid. The PV data is dynamic and can change quickly. A weather forecast could be used to try a prediction in advance.
* I prefer nighttime over daytime when no notable excess energy is to be expected, as during daytime, the charging can be interrupted when somebody needs the car.
* Bonus: Add calendar events where you know that the car can't be at the charging station as hard constraints.
* Bonus: Change target SoC based on calendar events when it can be inferred that you need a larger range because of an upcoming business trip or planned vacation.

I'm not looking for a concrete solution, but I want to get an idea of whether Timefold would be capable of modeling this optimization problem. The output should be the charging current (within a range supported by the wall box, like between 6A and 16A). Combining this with a solution like evcc.io would provide significant value over the existing smart charging solutions.

**Comment by triceo** [comment]
Hello @rhuss!

An interesting question. The answer, I think, would have to be based on what proportion of prediction is necessary here. If this is mostly prediction and just a little bit of scheduling, then Timefold Solver is not the right fit. We can not be used for predictions and projections.

However, if you can do the predictions in some other tool, and then give the solver a fixed list of costs at particular times, then I'd say Timefold Solver is a good fit. It's not necessarily a problem that the predictions would be changing on, say, an hourly basis - you can use [real-time planning](https://docs.timefold.ai/timefold-solver/latest/responding-to-change/responding-to-change#realTimePlanning) for that. As long as, while the solver is running, the costs are fixed.

**Reply by rhuss** [reply]
Thanks for the prompt answer. Indeed, prediction vs. scheduling is a good point. I agree that Timefold can be part of a solution, but you probably need more. Also, I'm not sure if the scheduling is "complex enough" to justify a sophisticated tool like Timefold.

Thanks also for the link to real-time planning; that might be very useful in this context. Disclaimer: I'm currently not working on this, but I was just curious if people have already thought about that domain.
